### ISSUE TITLE
[BUG] Cannot evict pod as it would violate the pod's disruption budget.

### ISSUE THREAD

**Issue by kobo-gcyre** [longhorn/longhorn#5910]
## Describe the bug (🐛 if you encounter this issue)

After upgrading RKE2 control nodes from v1.22.8+rke2r1 to v1.25.9+rke2r1 agent nodes cannot be drained and get stuck with
Cannot evict pod as it would violate the pod's disruption budget.
## To Reproduce

On RKE2 control node update to latest release and restart rke2
curl -sfL https://get.rke2.io | sh -
systemctl restart rke2-server

Perform steps on all controller nodes. Once upgrade is complete try draining agent node 
kubectl cordon agent-node
kubectl drain agent-node --ignore-daemonsets --delete-emptydir-data --pod-selector='app!=csi-attacher,app!=csi-provisioner,app!=longhorn-admission-webhook,app!=longhorn-conversion-webhook,app!=longhorn-driver-deployer' --grace-period=10

## Expected behavior
Agent node is drained.

## Log or Support bundle

If applicable, add the Longhorn managers' log or support bundle when the issue happens. 
You can generate a Support Bundle using the link at the footer of the Longhorn UI.

## Environment

 - Longhorn version: 1.26 (have also tried upgrading Longhorn to 1.33)
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: v1.22.8+rke2r1
   - Number of management node in the cluster: 3
   - Number of worker node in the cluster: 4
 - Node config
   - OS type and version: Ubuntu 20.04
   - CPU per node: 4 
   - Memory per node: 8GB
   - Disk type(e.g. SSD/NVMe): NVMe
   - Network bandwidth between the nodes: 10GB
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): VMWare
 - Number of Longhorn volumes in the cluster: 1

## Additional context
RKE2 upgrade is being done from 1.22 to 1.25 not in incremental steps (don't know if that has any impact or not)
From what I can tell the PDBs are not being deleted when the agent node is cordoned.

Pods and PDBs before cordon
```
csi-attacher-5f46994f7-9629c                             1/1     Running   0          20h   10.42.3.10  
csi-resizer-6dd8bd4c97-srr7d                             1/1     Running   0          20h   10.42.3.11
csi-snapshotter-86f65d8bc-p7p6q                          1/1     Running   0          20h   10.42.3.12
engine-image-ei-9bf563e8-wjm88                           1/1     Running   0          20h   10.42.3.7
instance-manager-e-6defff48                              1/1     Running   0          20h   10.42.3.8
instance-manager-r-d2f0328a                              1/1     Running   0          20h   10.42.3.9
longhorn-csi-plugin-rvhgz                                2/2     Running   0          20h   10.42.3.13
longhorn-manager-bx4x6                                   1/1     Running   0          20h   10.42.3.5
longhorn-ui-57c49478dc-s4vtl                             1/1     Running   0          20h   10.42.3.6
```

```
csi-attacher-5f46994f7-9629c                             1/1     Running   0          20h   10.42.3.10  
csi-resizer-6dd8bd4c97-srr7d                             1/1     Running   0          20h   10.42.3.11
csi-snapshotter-86f65d8bc-p7p6q                          1/1     Running   0          20h   10.42.3.12
engine-image-ei-9bf563e8-wjm88                           1/1     Running   0          20h   10.42.3.7
instance-manager-e-6defff48                              1/1     Running   0          20h   10.42.3.8
instance-manager-r-d2f0328a                              1/1     Running   0          20h   10.42.3.9
longhorn-csi-plugin-rvhgz                                2/2     Running   0          20h   10.42.3.13
longhorn-manager-bx4x6                                   1/1     Running   0          20h   10.42.3.5
longhorn-ui-57c49478dc-s4vtl                             1/1     Running   0          20h   10.42.3.6
```

after cordon
```
kubectl cordon agent-node
node/agent-node cordoned

kubectl -n longhorn-system get pdb
NAME                          MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
instance-manager-e-17712f99   1               N/A               0                     20h
instance-manager-e-6defff48   1               N/A               0                     20h
instance-manager-e-9c26f77c   1               N/A               0                     20h
instance-manager-e-d82e16a5   1               N/A               0                     20h
instance-manager-r-23246e55   1               N/A               0                     20h
instance-manager-r-37ca8394   1               N/A               0                     20h
instance-manager-r-d2f0328a   1               N/A               0                     20h
instance-manager-r-e3eda6dd   1               N/A               0                     20h
```

[longhorn-support-bundle_6c16098f-a97d-41e6-8aa8-1329a6535eb8_2023-05-10T16-57-07Z.zip](https://github.com/longhorn/longhorn/files/11444384/longhorn-support-bundle_6c16098f-a97d-41e6-8aa8-1329a6535eb8_2023-05-10T16-57-07Z.zip)

**Comment by kobo-gcyre** [comment]
trying to drain the node
```kubectl drain agent-node --ignore-daemonsets --delete-emptydir-data --pod-selector='app!=csi-attacher,app!=csi-provisioner,app!=longhorn-admission-webhook,app!=longhorn-conversion-webhook,app!=longhorn-driver-deployer' --grace-period=10
node/agent-node already cordoned
Warning: ignoring DaemonSet-managed Pods: kube-system/rke2-canal-kmldj, kube-system/rke2-ingress-nginx-controller-xmt55, longhorn-system/engine-image-ei-9bf563e8-wjm88, longhorn-system/longhorn-csi-plugin-rvhgz, longhorn-system/longhorn-manager-bx4x6
evicting pod longhorn-system/csi-resizer-6dd8bd4c97-srr7d
evicting pod static-test/kobo-static-deployment-5b587b77cb-xxg8t
evicting pod longhorn-system/instance-manager-r-d2f0328a
evicting pod longhorn-system/longhorn-ui-57c49478dc-s4vtl
evicting pod longhorn-system/csi-snapshotter-86f65d8bc-p7p6q
evicting pod longhorn-system/instance-manager-e-6defff48
evicting pod kube-system/helm-install-rke2-snapshot-controller-8sb94
evicting pod kube-system/rke2-coredns-rke2-coredns-autoscaler-5c6c4bc849-gwqhc
evicting pod static-test/kobo-static-deployment-5b587b77cb-bxdpm
error when evicting pods/"instance-manager-r-d2f0328a" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pods/"instance-manager-e-6defff48" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
pod/helm-install-rke2-snapshot-controller-8sb94 evicted
pod/csi-resizer-6dd8bd4c97-srr7d evicted
pod/rke2-coredns-rke2-coredns-autoscaler-5c6c4bc849-gwqhc evicted
pod/kobo-static-deployment-5b587b77cb-xxg8t evicted
pod/csi-snapshotter-86f65d8bc-p7p6q evicted
pod/kobo-static-deployment-5b587b77cb-bxdpm evicted
evicting pod longhorn-system/instance-manager-r-d2f0328a
evicting pod longhorn-system/instance-manager-e-6defff48
error when evicting pods/"instance-manager-e-6defff48" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pods/"instance-manager-r-d2f0328a" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.```

**Comment by ChanYiLin** [comment]
Hi @kobo-gcyre , we have fixed some pdb issue after Longhorn v.1.3.3
https://github.com/longhorn/longhorn/issues/5680
Maybe let's upgrade the longhorn to v1.3.3 and try to see if this issue still happens?
Thanks

**Comment by kobo-gcyre** [comment]
I ran another test this morning but still seeing the issue. This time the cluster was upgraded to from v1.22.9+rke2r2 to v1.24.13+rke2r1 then longhorn upgraded to 1.33

Steps:
Cluster 3 control nodes, 4 worker nodes 
RKE2 version: v1.22.9+rke2r2
Longhorn version: 1.26

Upgrade 3 control nodes to v1.23.17+rke2r1
Upgrade 4 worker nodes to v1.23.17+rke2r1

Upgrade 3 control nodes to v1.24.13+rke2r1
Upgrade 4 worker nodes to v1.24.13+rke2r1
Upgrade longhorn to 1.33

At this point everything is working as expected, nodes can be cordoned and drained

Upgrade 3 control nodes to v1.25.9+rke2r1 

cordon worker node for upgrade
drain node, 
```

evicting pod longhorn-system/instance-manager-r-f027b639202c7a45a026f2f39622dd4a
evicting pod longhorn-system/instance-manager-e-f027b639202c7a45a026f2f39622dd4a
error when evicting pods/"instance-manager-e-f027b639202c7a45a026f2f39622dd4a" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
error when evicting pods/"instance-manager-r-f027b639202c7a45a026f2f39622dd4a" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

**Comment by kobo-gcyre** [comment]
@ChanYiLin any updates?

**Comment by ChanYiLin** [comment]
Sorry for the late update.
The issue is because longhorn try to temporarily delete the pdb during the upgrade, but somehow failed to deleted it, and here is the reason.
May need to take a look.

```
2023-05-12T13:36:49.304438231Z W0512 13:36:49.304315       1 reflector.go:324] k8s.io/client-go/informers/factory.go:134: failed to list *v1beta1.PodDisruptionBudget: the server could not find the requested resource
2023-05-12T13:36:49.304481631Z E0512 13:36:49.304351       1 reflector.go:138] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1beta1.PodDisruptionBudget: failed to list *v1beta1.PodDisruptionBudget: the server could not find the requested resource
2023-05-12T13:37:15.770499856Z time="2023-05-12T13:37:15Z" level=debug msg="Node hd-k8swrkr-031 is marked unschedulable, try to remove instance-manager-e-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
2023-05-12T13:37:15.771080734Z time="2023-05-12T13:37:15Z" level=debug msg="Node hd-k8swrkr-031 is marked unschedulable, try to remove instance-manager-r-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
2023-05-12T13:37:15.772373509Z time="2023-05-12T13:37:15Z" level=info msg="Deleted instance-manager-e-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
2023-05-12T13:37:15.772648227Z time="2023-05-12T13:37:15Z" level=info msg="Deleted instance-manager-r-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
2023-05-12T13:37:28.984644220Z W0512 13:37:28.984415       1 reflector.go:324] k8s.io/client-go/informers/factory.go:134: failed to list *v1beta1.CronJob: the server could not find the requested resource
2023-05-12T13:37:28.984735007Z E0512 13:37:28.984457       1 reflector.go:138] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1beta1.CronJob: failed to list *v1beta1.CronJob: the server could not find the requested resource
2023-05-12T13:37:39.833991100Z W0512 13:37:39.833843       1 reflector.go:324] k8s.io/client-go/informers/factory.go:134: failed to list *v1beta1.PodDisruptionBudget: the server could not find the requested resource
2023-05-12T13:37:39.834055088Z E0512 13:37:39.833976       1 reflector.go:138] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1beta1.PodDisruptionBudget: failed to list *v1beta1.PodDisruptionBudget: the server could not find the requested resource
2023-05-12T13:37:45.771543528Z time="2023-05-12T13:37:45Z" level=debug msg="Node hd-k8swrkr-031 is marked unschedulable, try to remove instance-manager-e-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
2023-05-12T13:37:45.771626303Z time="2023-05-12T13:37:45Z" level=debug msg="Node hd-k8swrkr-031 is marked unschedulable, try to remove instance-manager-r-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
2023-05-12T13:37:45.773089095Z time="2023-05-12T13:37:45Z" level=info msg="Deleted instance-manager-r-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
2023-05-12T13:37:45.773118092Z time="2023-05-12T13:37:45Z" level=info msg="Deleted instance-manager-e-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
```

cc @PhanLe1010

**Comment by ChanYiLin** [comment]
I just found that PodDisruptionBudgets API version policy/v1beta1 is removed in Kubernetes 1.25
https://github.com/kubernetes-sigs/metrics-server/issues/1104

This can also be seen in the manager log
```
2023-05-12T12:53:54.990466212Z W0512 12:53:54.990228       1 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
2023-05-12T12:53:54.990488219Z time="2023-05-12T12:53:54Z" level=info msg="Created instance-manager-r-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
2023-05-12T12:53:55.009319680Z time="2023-05-12T12:53:55Z" level=debug msg="Start monitoring instance manager instance-manager-r-f027b639202c7a45a026f2f39622dd4a" controller=longhorn-instance-manager instance manager=instance-manager-r-f027b639202c7a45a026f2f39622dd4a node=hd-k8swrkr-031
2023-05-12T12:53:55.022466758Z time="2023-05-12T12:53:55Z" level=info msg="Instance manager state is updated from starting to running after syncing with the pod" controller=longhorn-instance-manager instanceManager=instance-manager-e-f027b639202c7a45a026f2f39622dd4a node=hd-k8swrkr-031 nodeID=hd-k8swrkr-031
2023-05-12T12:53:55.046893276Z W0512 12:53:55.046755       1 warnings.go:70] policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget
2023-05-12T12:53:55.047374140Z time="2023-05-12T12:53:55Z" level=info msg="Created instance-manager-e-f027b639202c7a45a026f2f39622dd4a PDB" controller=longhorn-instance-manager node=hd-k8swrkr-031
```
Can you help me to check your Kubernetes version
According to our doc, v1.3.3 has constraint for Kubernetes version `Kubernetes >= v1.18 and <= v1.24`
https://longhorn.io/docs/1.3.3/deploy/install/#installation-requirements

thanks!

**Comment by aleksasiriski** [comment]
This still happens on v1.4.2

**Comment by gcarrarom** [comment]
Still happening on 1.5.0

**Comment by PhanLe1010** [comment]
@gcarrarom Could you give the reproduce steps?

**Comment by gcarrarom** [comment]
It is unfortunately not consistent. It happens sometimes as I drain the nodes for patches. The node just started reporting that it couldn't evict the pod for hours until I saw it. There are no disks with less than 3 replicas, and all nodes have enough space for the replicas if two nodes fail.

**Comment by PhanLe1010** [comment]
When it happens next time, could you send us a support bundle to `longhorn-support-bundle@Suse.com` taken while the issue is happening? Thank you

**Comment by emmeowzing** [comment]
This is happening on one of my clusters as well. Same versions.

**Comment by emmeowzing** [comment]
I found just now that if I deleted the proper pdb in a separate shell, drain would complete on its own.

**Comment by schmidp** [comment]
I have the same problem:

(the longhorn-manager has such a low age because I tried killing it)

<img width="1823" alt="Screenshot 2023-09-05 at 22 28 39" src="https://github.com/longhorn/longhorn/assets/10045/fef4b3c0-0a8d-42b0-ad66-9a0e3c92db6a">

<img width="1823" alt="Screenshot 2023-09-05 at 22 28 48" src="https://github.com/longhorn/longhorn/assets/10045/2514f697-cc26-42bb-90c7-3078b00b5b39">

to fix it I manually modified the pod disruption budgets.

I noted that the for the nodes where it worked the pod disruption budget was recently created (age was about 30 minutes old -> from when the node was updated). for the nodes where it got stuck the PDB was 30 days old (probably when I installed longhorn).

**Comment by aleksasiriski** [comment]
Manually deleting PDB made the node successfully drain:
[supportbundle_988f1b52-5480-4f4c-b162-a08a9fbdd39c_2023-09-22T11-31-20Z.zip](https://github.com/longhorn/longhorn/files/12700244/supportbundle_988f1b52-5480-4f4c-b162-a08a9fbdd39c_2023-09-22T11-31-20Z.zip)

**Comment by HendrikLevering** [comment]
I get a similar error with longhorn 1.5.1 and strict local nodes. The pdb is:
```apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  creationTimestamp: "2023-09-19T11:36:36Z"
  generation: 1
  name: instance-manager-266f4b81cf5ad313fd2c15b1931065d0
  namespace: longhorn-system
  resourceVersion: "9563034"
  uid: a6e0bdea-22e0-4da0-8da2-60eed9d3db1c
spec:
  minAvailable: 1
  selector:
    matchLabels:
      longhorn.io/component: instance-manager
      longhorn.io/instance-manager-image: imi-a21ea71c
      longhorn.io/instance-manager-type: aio
      longhorn.io/managed-by: longhorn-manager
      longhorn.io/node: k3s-agent-small-postgres-vfo
status:
  conditions:
  - lastTransitionTime: "2023-09-23T08:59:40Z"
    message: ""
    observedGeneration: 1
    reason: InsufficientPods
    status: "False"
    type: DisruptionAllowed
  currentHealthy: 1
  desiredHealthy: 1
  disruptionsAllowed: 0
  expectedPods: 1
  observedGeneration: 1```

It turns out that iscsid is not running. It fails to start with this error 
`Oct 04 11:54:55 k3s-agent-small-postgres-gbt systemd[1]: iscsid.service: Job iscsid.service/start failed with result 'dependency'.
Oct 04 11:54:55 k3s-agent-small-postgres-gbt systemd[1]: iscsid.socket: Trigger limit hit, refusing further activation.
Oct 04 11:54:55 k3s-agent-small-postgres-gbt systemd[1]: iscsid.socket: Failed with result 'trigger-limit-hit'.`

**Comment by HendrikLevering** [comment]
I have the same issue with longhorn 1.5.1 and strict local storage. node_drain_policy is allow_if_replica_is_stopped. The volumes are detached. But the error appears anyway

**Comment by janosmiko** [comment]
I face the same issue even when actually Longhorn Frontend shows that I have 0 replicas on that node (but the pod disruption budget is set to have at least 1 replica of the instance-manager-xyz)

**Comment by simonebenati** [comment]
Facing the same issue currently longhorn v1.4.3 - What can we do to help you debug the issue?

**Comment by fmiqbal** [comment]
Encountering this issue today on longhorn v1.5.3, with microk8s 1.28.3
its not exactly "a bug" per se in my case, its just that it needs more meaningful error message or something to troubleshoot, i got 2 different case on 2 different node from a 5 node cluster.

#### its because there are strict-local node

this maybe obvious, but for my step, first I kubectl drain, and then it stuck on pdb, and then I got 2nd pod of my 2-pod cnpg cluster that have been evicted on that node, cannot be started because it has strict local and the replica doesnt moved by longhorn (?, although, i didnt know if this is intended behaviour or not), the replica show as detached in longhorn

after that I also verify that there are no more strict local or another last single replica on the drained node, this is a bit hard because I cant filter by replica count nor by the data locality, the page showing only 100 per page, although for the data locality i can use browser ctrl-f, the replica count i need to scroll slowly through the table

#### `somehow™` a replica is failing on a non started pod (detached volume)

it doesnt show much when its on detached, but when I start the pod, the replica show as failing/failed/red, so i delete that replica, and now it can drain successfully while the replica created on another schedulable node

So, I think we need a way to fastly debug/find the problematic replica rather than to check by each case,  or maybe show it in the instance manager log what replica is blocking the drain

**Comment by innobead** [comment]
cc @james-munson

**Comment by bananflugan** [comment]
This started happening to me last week when i upgraded from longhorn 1.5.1 -> 1.5.4.
The RKE clusters are running 1.26.8.

Draining procedure gets stuck at:
`error when evicting pods/"instance-manager-992d684c4a70f8d17d28be3fe762bf61" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.`

This is very frustrating as my drain and patch Ansible playbook cant be used anymore.

**Comment by gerwim** [comment]
I have the exact same issue, but I'm running the [system upgrade controller](https://github.com/rancher/system-upgrade-controller).

Some logs:

```
| stream logs failed container "upgrade" in pod "apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q" is waiting to start: PodInitializing for system-upgrade/apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q (upgrade)                                                                                                                      │
│ stream logs failed container "upgrade" in pod "apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q" is waiting to start: PodInitializing for system-upgrade/apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q (upgrade)                                                                                                                      │
│ stream logs failed container "upgrade" in pod "apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q" is waiting to start: PodInitializing for system-upgrade/apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q (upgrade)                                                                                                                      │
│ drain evicting pod longhorn-system/instance-manager-a20036ac19c45e8de43a4b00c7ef7153                                                                                                                                                                                                                                                                                   │
│ drain error when evicting pods/"instance-manager-a20036ac19c45e8de43a4b00c7ef7153" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.                                                                                                                                                                       │
│ stream logs failed container "upgrade" in pod "apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q" is waiting to start: PodInitializing for system-upgrade/apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q (upgrade)                                                                                                                      │
│ stream logs failed container "upgrade" in pod "apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q" is waiting to start: PodInitializing for system-upgrade/apply-k3s-updater-on-k3s-2-with-8a092280f0e00a572a9bf1f73-9hb8q (upgrade)                                                                                                                      │
│ drain evicting pod longhorn-system/instance-manager-a20036ac19c45e8de43a4b00c7ef7153                                                                                                                                                                                                                                                                                   │
│ drain error when evicting pods/"instance-manager-a20036ac19c45e8de43a4b00c7ef7153" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.                                                                                                                                                                       │
│ drain evicting pod longhorn-system/instance-manager-a20036ac19c45e8de43a4b00c7ef7153
```

```
➜  ~ k -n longhorn-system get pdb
NAME                                                MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
csi-attacher                                        1               N/A               2                     4m55s
csi-provisioner                                     1               N/A               2                     4m55s
instance-manager-702404b9821a6e7eb430b5f9540b75ce   1               N/A               0                     4m19s
instance-manager-a20036ac19c45e8de43a4b00c7ef7153   1               N/A               0                     6m12s
instance-manager-f3ae94951312e170de59a63529198987   1               N/A               0                     5m42s
```

To upgrade, I need to manually remove the PDB's to make it work. This prevents me from using the auto update feature, since the node will keep the status `SchedulingDisabled`.

**Comment by Eerosal** [comment]
We are also encountering this issue with node draining due to Longhorn PDBs in our setup. We suspect that it is causing issues with our autoscaler setup. The Longhorn configuration is set so that the system is aware of the autoscaler.

`evicting pod longhorn-system/instance-manager-d251a5e6222543fb04d53f348f406435
error when evicting pods/"instance-manager-d251a5e6222543fb04d53f348f406435" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.`

Helm chart version 1.6.1

**Comment by Littlericket** [comment]
Atleast for 1.6.1 it seems to have been fixed.
When upgrading worker nodes, the PDB gets deleted by longhorn-manager when all remaining pods having PVs are detached. Still, this only works when using the new introduced setting from https://github.com/longhorn/longhorn/issues/7421 - set either to "block-for-eviction" or "block-for-eviction-if-contains-last-replica". If all replicas are evicted, the PDB gets removed by the longhorn-manager.

**Comment by gerwim** [comment]
To add on @Littlericket's comment. I have just upgraded my 3 node cluster to 1.28, and it requires a 2 replica count of the volumes. I had it on three (because I have a three node cluster), but this prevents the`system-upgrade-controller` to complete the upgrades, since the PDB blocks the eviction.

**Comment by kastl-ars** [comment]
I am still encountering the issue on a three-node k3s cluster (with all nodes being workers and controlplane).

I tried the different settings for `NodeDrainPolicy`, but they did not improve the situation. (Or maybe they are not respected when changing this parameter after the initial deployment).

```
evicting pod longhorn-system/instance-manager-5131d31d0f1e2242c9793a306c6b404b
error when evicting pods/"instance-manager-5131d31d0f1e2242c9793a306c6b404b" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

Currently on Longhorn 1.6.2 installed via ArgoCD.

**Comment by bananflugan** [comment]
> I am still encountering the issue on a three-node k3s cluster (with all nodes being workers and controlplane).
> 
> I tried the different settings for `NodeDrainPolicy`, but they did not improve the situation. (Or maybe they are not respected when changing this parameter after the initial deployment).
> 
> ```
> evicting pod longhorn-system/instance-manager-5131d31d0f1e2242c9793a306c6b404b
> error when evicting pods/"instance-manager-5131d31d0f1e2242c9793a306c6b404b" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
> ```
> 
> Currently on Longhorn 1.6.2 installed via ArgoCD.

Make sure you have 1 more longhorn node than replicas configured. The last replica must have somewhere to go, otherwise it cant be evicted. So if you have longhorn configured to have 3 replicas you must have 4 nodes for longhorn.

I had the same problem before with running 3 replicas and only 3 nodes which worked before but not anymore.

**Comment by kastl-ars** [comment]
> Make sure you have 1 more longhorn node than replicas configured. The last replica must have somewhere to go, otherwise it cant be evicted. So if you have longhorn configured to have 3 replicas you must have 4 nodes for longhorn.
> 
> I had the same problem before with running 3 replicas and only 3 nodes which worked before but not anymore.

Thanks, I'll give this a try.

**Comment by w13915984028** [comment]
This issue can be reproduced on Harvester, one course is:

[[BUG] A manual attached LH volume breaks Harvester maintenance mode](https://github.com/harvester/harvester/issues/6264#top)

For anyone who encounters this error, please check below CRD object, if any is from `longhorn-ui`, then LH will related pod is failed to be evicted
```
time="2024-08-05T09:49:11Z" level=info msg="error when evicting pods/\"instance-manager-68cd2514dd3f6d59b95cbd865d5b08f7\" -n \"longhorn-system\" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget."
```

```
- apiVersion: longhorn.io/v1beta2
  kind: VolumeAttachment
...
  spec:
    attachmentTickets:
      longhorn-ui:
        generation: 0
        id: longhorn-ui
        nodeID: s37570031
        parameters:
          disableFrontend: "false"
          lastAttachedBy: "null"
        type: longhorn-api
    volume: pvc-9b35136c-f59e-414b-aa55-b84b9b21ff89
```

**Comment by w13915984028** [comment]
Another cause is also observed: 

>When a  volume is created from a single-replica storageclass, even after the volume is detached from the target node, LH still blocks the node drain.

The new debug message:

>time="2024-08-05T15:07:41Z" level=info msg="Node instance-manager-68cd2514dd3f6d59b95cbd865d5b08f7 is marked unschedulable but removing harv2 PDB is blocked: no pdb on another node, wait " func="controller.(*InstanceManagerController).syncInstanceManagerPDB" file="instance_manager_controller.go:730" controller=longhorn-instance-manager node=harv2

For single-replica volume,  can't expect there is at least an another pdb on other nodes.

A new PR will be added to fix this.

**Comment by derekbit** [comment]
Hello @ChanYiLin, can you help check it? Thank you.

**Comment by AntonioCayulao** [comment]
Hi all,
we are using Longhorn v1.7.0 helm chart version and we have the same issue.
Our system is related to IoT features so we have single-node clusters spread around the world. These edge devices constantly are turned off/on as a every day operation.

In the past we didn't notice this issue till now that we found that another component GPU-Operator from Nvidia do the drain/cordon operation to deploy their pods. Every time this happens we need to manually go inside the host to uncordon the node.

I noticed that the process was failing because longhorn was blocking the drain process with this error message:
```bash
evicting pod longhorn-system/csi-attacher-8895d78cb-r7cn8
error when evicting pods/"csi-attacher-8895d78cb-r7cn8" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```
So I modified the `nodeDrainPolicy` to `allow-if-replica-is-stopped` and now every time we restarted the host the cordon operation happens and even when we manually uncordon the node longhorn is cordoning the node again.

We don't know how to fix it.
This is our values.yaml config regardless to the default config:
```yaml
defaultSettings:
  # -- Endpoint used to access the backupstore. (Options: "NFS", "CIFS", "AWS", "GCP", "AZURE")
  backupTarget: ~
  # -- Name of the Kubernetes secret associated with the backup target.
  backupTargetCredentialSecret: ~
  # -- Setting that allows Longhorn to automatically attach a volume and create snapshots or backups when recurring jobs are run.
  allowRecurringJobWhileVolumeDetached: ~
  # -- Setting that allows Longhorn to automatically create a default disk only on nodes with the label "node.longhorn.io/create-default-disk=true" (if no other disks exist). When this setting is disabled, Longhorn creates a default disk on each node that is added to the cluster.
  createDefaultDiskLabeledNodes: ~
  # -- Default path for storing data on a host. The default value is "/var/lib/longhorn/".
  defaultDataPath: ~
  # -- Default data locality. A Longhorn volume has data locality if a local replica of the volume exists on the same node as the pod that is using the volume.
  defaultDataLocality: "strict-local"
  # -- Setting that allows scheduling on nodes with healthy replicas of the same volume. This setting is disabled by default.
  replicaSoftAntiAffinity: false
  # -- Setting that automatically rebalances replicas when an available node is discovered.
  replicaAutoBalance: ~
  # -- Percentage of storage that can be allocated relative to hard drive capacity. The default value is "100".
  storageOverProvisioningPercentage: ~
  # -- Percentage of minimum available disk capacity. When the minimum available capacity exceeds the total available capacity, the disk becomes unschedulable until more space is made available for use. The default value is "25".
  storageMinimalAvailablePercentage: 10
  # -- Percentage of disk space that is not allocated to the default disk on each new Longhorn node.
  storageReservedPercentageForDefaultDisk: 10
  # -- Upgrade Checker that periodically checks for new Longhorn versions. When a new version is available, a notification appears on the Longhorn UI. This setting is enabled by default
  upgradeChecker: false
  # -- Default number of replicas for volumes created using the Longhorn UI. For Kubernetes configuration, modify the `numberOfReplicas` field in the StorageClass. The default value is "3".
  defaultReplicaCount: 1
  # -- Default Longhorn StorageClass. "storageClassName" is assigned to PVs and PVCs that are created for an existing Longhorn volume. "storageClassName" can also be used as a label, so it is possible to use a Longhorn StorageClass to bind a workload to an existing PV without creating a Kubernetes StorageClass object. The default value is "longhorn-static".
  defaultLonghornStaticStorageClass: ~
  # -- Number of seconds that Longhorn waits before checking the backupstore for new backups. The default value is "300". When the value is "0", polling is disabled.
  backupstorePollInterval: ~
  # -- Number of minutes that Longhorn keeps a failed backup resource. When the value is "0", automatic deletion is disabled.
  failedBackupTTL: ~
  # -- Setting that restores recurring jobs from a backup volume on a backup target and creates recurring jobs if none exist during backup restoration.
  restoreVolumeRecurringJobs: ~
  # -- Maximum number of successful recurring backup and snapshot jobs to be retained. When the value is "0", a history of successful recurring jobs is not retained.
  recurringSuccessfulJobsHistoryLimit: ~
  # -- Maximum number of failed recurring backup and snapshot jobs to be retained. When the value is "0", a history of failed recurring jobs is not retained.
  recurringFailedJobsHistoryLimit: 1
  # -- Maximum number of snapshots or backups to be retained.
  recurringJobMaxRetention: ~
  # -- Maximum number of failed support bundles that can exist in the cluster. When the value is "0", Longhorn automatically purges all failed support bundles.
  supportBundleFailedHistoryLimit: ~
  # -- Taint or toleration for system-managed Longhorn components.
  # Specify values using a semicolon-separated list in `kubectl taint` syntax (Example: key1=value1:effect; key2=value2:effect).
  taintToleration: ~
  # -- Node selector for system-managed Longhorn components.
  systemManagedComponentsNodeSelector: ~
  # -- PriorityClass for system-managed Longhorn components.
  # This setting can help prevent Longhorn components from being evicted under Node Pressure.
  # Notice that this will be applied to Longhorn user-deployed components by default if there are no priority class values set yet, such as `longhornManager.priorityClass`.
  priorityClass: &defaultPriorityClassNameRef "longhorn-critical"
  # -- Setting that allows Longhorn to automatically salvage volumes when all replicas become faulty (for example, when the network connection is interrupted). Longhorn determines which replicas are usable and then uses these replicas for the volume. This setting is enabled by default.
  autoSalvage: ~
  # -- Setting that allows Longhorn to automatically delete a workload pod that is managed by a controller (for example, daemonset) whenever a Longhorn volume is detached unexpectedly (for example, during Kubernetes upgrades). After deletion, the controller restarts the pod and then Kubernetes handles volume reattachment and remounting.
  autoDeletePodWhenVolumeDetachedUnexpectedly: true
  # -- Setting that prevents Longhorn Manager from scheduling replicas on a cordoned Kubernetes node. This setting is enabled by default.
  disableSchedulingOnCordonedNode: true
  # -- Setting that allows Longhorn to schedule new replicas of a volume to nodes in the same zone as existing healthy replicas. Nodes that do not belong to any zone are treated as existing in the zone that contains healthy replicas. When identifying zones, Longhorn relies on the label "topology.kubernetes.io/zone=<Zone name of the node>" in the Kubernetes node object.
  replicaZoneSoftAntiAffinity: false
  # -- Setting that allows scheduling on disks with existing healthy replicas of the same volume. This setting is enabled by default.
  replicaDiskSoftAntiAffinity: ~
  # -- Policy that defines the action Longhorn takes when a volume is stuck with a StatefulSet or Deployment pod on a node that failed.
  nodeDownPodDeletionPolicy: "delete-both-statefulset-and-deployment-pod"
  # -- Policy that defines the action Longhorn takes when a node with the last healthy replica of a volume is drained.
  nodeDrainPolicy: "allow-if-replica-is-stopped"
  # -- Setting that allows automatic detaching of manually-attached volumes when a node is cordoned.
  detachManuallyAttachedVolumesWhenCordoned: true
  # -- Number of seconds that Longhorn waits before reusing existing data on a failed replica instead of creating a new replica of a degraded volume.
  replicaReplenishmentWaitInterval: ~
  # -- Maximum number of replicas that can be concurrently rebuilt on each node.
  concurrentReplicaRebuildPerNodeLimit: ~
  # -- Maximum number of volumes that can be concurrently restored on each node using a backup. When the value is "0", restoration of volumes using a backup is disabled.
  concurrentVolumeBackupRestorePerNodeLimit: ~
  # -- Setting that disables the revision counter and thereby prevents Longhorn from tracking all write operations to a volume. When salvaging a volume, Longhorn uses properties of the "volume-head-xxx.img" file (the last file size and the last time the file was modified) to select the replica to be used for volume recovery. This setting applies only to volumes created using the Longhorn UI.
  disableRevisionCounter: "true"
  # -- Image pull policy for system-managed pods, such as Instance Manager, engine images, and CSI Driver. Changes to the image pull policy are applied only after the system-managed pods restart.
  systemManagedPodsImagePullPolicy: "if-not-present"
  # -- Setting that allows you to create and attach a volume without having all replicas scheduled at the time of creation.
  allowVolumeCreationWithDegradedAvailability: false
  # -- Setting that allows Longhorn to automatically clean up the system-generated snapshot after replica rebuilding is completed.
  autoCleanupSystemGeneratedSnapshot: true
  # -- Setting that allows Longhorn to automatically clean up the snapshot generated by a recurring backup job.
  autoCleanupRecurringJobBackupSnapshot: true
  # -- Maximum number of engines that are allowed to concurrently upgrade on each node after Longhorn Manager is upgraded. When the value is "0", Longhorn does not automatically upgrade volume engines to the new default engine image version.
  concurrentAutomaticEngineUpgradePerNodeLimit: 3
  # -- Number of minutes that Longhorn waits before cleaning up the backing image file when no replicas in the disk are using it.
  backingImageCleanupWaitInterval: ~
  # -- Number of seconds that Longhorn waits before downloading a backing image file again when the status of all image disk files changes to "failed" or "unknown".
  backingImageRecoveryWaitInterval: ~
  # -- Percentage of the total allocatable CPU resources on each node to be reserved for each instance manager pod when the V1 Data Engine is enabled. The default value is "12".
  guaranteedInstanceManagerCPU: 5
  # -- Setting that notifies Longhorn that the cluster is using the Kubernetes Cluster Autoscaler.
  kubernetesClusterAutoscalerEnabled: ~
  # -- Setting that allows Longhorn to automatically delete an orphaned resource and the corresponding data (for example, stale replicas). Orphaned resources on failed or unknown nodes are not automatically cleaned up.
  orphanAutoDeletion: true
  # -- Storage network for in-cluster traffic. When unspecified, Longhorn uses the Kubernetes cluster network.
  storageNetwork: ~
  # -- Flag that prevents accidental uninstallation of Longhorn.
  deletingConfirmationFlag: ~
  # -- Timeout between the Longhorn Engine and replicas. Specify a value between "8" and "30" seconds. The default value is "8".
  engineReplicaTimeout: ~
  # -- Setting that allows you to enable and disable snapshot hashing and data integrity checks.
  snapshotDataIntegrity: ~
  # -- Setting that allows disabling of snapshot hashing after snapshot creation to minimize impact on system performance.
  snapshotDataIntegrityImmediateCheckAfterSnapshotCreation: ~
  # -- Setting that defines when Longhorn checks the integrity of data in snapshot disk files. You must use the Unix cron expression format.
  snapshotDataIntegrityCronjob: ~
  # -- Setting that allows Longhorn to automatically mark the latest snapshot and its parent files as removed during a filesystem trim. Longhorn does not remove snapshots containing multiple child files.
  removeSnapshotsDuringFilesystemTrim: ~
  # -- Setting that allows fast rebuilding of replicas using the checksum of snapshot disk files. Before enabling this setting, you must set the snapshot-data-integrity value to "enable" or "fast-check".
  fastReplicaRebuildEnabled: ~
  # -- Number of seconds that an HTTP client waits for a response from a File Sync server before considering the connection to have failed.
  replicaFileSyncHttpClientTimeout: ~
  # -- Number of seconds that Longhorn allows for the completion of replica rebuilding and snapshot cloning operations.
  longGRPCTimeOut: ~
  # -- Log levels that indicate the type and severity of logs in Longhorn Manager. The default value is "Info". (Options: "Panic", "Fatal", "Error", "Warn", "Info", "Debug", "Trace")
  logLevel: ~
  # -- Setting that allows you to specify a backup compression method.
  backupCompressionMethod: ~
  # -- Maximum number of worker threads that can concurrently run for each backup.
  backupConcurrentLimit: ~
  # -- Maximum number of worker threads that can concurrently run for each restore operation.
  restoreConcurrentLimit: ~
  # -- Setting that allows you to enable the V1 Data Engine.
  v1DataEngine: ~
  # -- Setting that allows you to enable the V2 Data Engine, which is based on the Storage Performance Development Kit (SPDK). The V2 Data Engine is a preview feature and should not be used in production environments.
  v2DataEngine: ~
  # -- Setting that allows you to configure maximum huge page size (in MiB) for the V2 Data Engine.
  v2DataEngineHugepageLimit: ~
  # -- Number of millicpus on each node to be reserved for each Instance Manager pod when the V2 Data Engine is enabled. The default value is "1250".
  v2DataEngineGuaranteedInstanceManagerCPU: ~
  # -- Setting that allows scheduling of empty node selector volumes to any node.
  allowEmptyNodeSelectorVolume: ~
  # -- Setting that allows scheduling of empty disk selector volumes to any disk.
  allowEmptyDiskSelectorVolume: ~
  # -- Setting that allows Longhorn to periodically collect anonymous usage data for product improvement purposes. Longhorn sends collected data to the [Upgrade Responder](https://github.com/longhorn/upgrade-responder) server, which is the data source of the Longhorn Public Metrics Dashboard (https://metrics.longhorn.io). The Upgrade Responder server does not store data that can be used to identify clients, including IP addresses.
  allowCollectingLonghornUsageMetrics: ~
  # -- Setting that temporarily prevents all attempts to purge volume snapshots.
  disableSnapshotPurge: ~
  # -- Maximum snapshot count for a volume. The value should be between 2 to 250
  snapshotMaxCount: ~
  # -- Setting that allows you to configure the log level of the SPDK target daemon (spdk_tgt) of the V2 Data Engine.
  v2DataEngineLogLevel: ~
  # -- Setting that allows you to configure the log flags of the SPDK target daemon (spdk_tgt) of the V2 Data Engine.
  v2DataEngineLogFlags: ~
  # -- Setting that freezes the filesystem on the root partition before a snapshot is created.
  freezeFilesystemForSnapshot: ~
  # -- Setting that automatically cleans up the snapshot when the backup is deleted.
  autoCleanupSnapshotWhenDeleteBackup: true
  # -- Turn on logic to detect and move RWX volumes quickly on node failure.
  rwxVolumeFastFailover: ~
```

**Comment by ChanYiLin** [comment]
Hi @AntonioCayulao 
Thanks for your report
Longhorn won't cordon the k8s node itself, it only reacts after the node is cordoned.
Would you help to collect the support bundle when the issue happens and email to us so we can further investigate the problem.
Thanks

**Comment by pat-s** [comment]
I am also frequently running into this, both for k3s updates and node OS reboots via `kured`.

I have three nodes and a full healthy replica set, so the blocking is not because the node in question contains the last replica of a volume.

I always have to manually delete the `pdb` to get the flow to continue. Reading this issue I can't really infer what determines if the `pdb` is set to 0 and what one could do to prevent manual action here.

For a cluster with multiple nodes and automatic reboots for OS updates and k3s updates, this is a nightmare.

**Comment by kastl-ars** [comment]
> I have two replicas across 3 nodes. So when being in a healthy state for all volumes, I'd expect that cordoning and then draining a node would work as there is at least one healthy replica remaining at all times.
> 
> When draining, I don't see replicas being moved by `longhorn`. They are rebuilt on another node when the node is finally gone, yes, but they are not moved to a different node before removing the node (which is OK - and I guess expected? - if it is not the last replica). Only when going into the UI and selecting "evict node", replicas are force moved.
> 
> I don't understand why the `pdb` is blocking actions if the replica situation seems actually fine. But given that only a few people report this here, I it likely that it is a configuration issue. I just don't see which one (yet).

Exactly this is what I see and what I do not understand...

**Comment by gcarrarom** [comment]
> I am also frequently running into this, both for k3s updates and node OS reboots via `kured`.
> 
> I have three nodes and a full healthy replica set, so the blocking is not because the node in question contains the last replica of a volume.
> 
> I always have to manually delete the `pdb` to get the flow to continue. Reading this issue I can't really infer what determines if the `pdb` is set to 0 and what one could do to prevent manual action here.
> 
> For a cluster with multiple nodes and automatic reboots for OS updates and k3s updates, this is a nightmare.
> 
> (v1.7.2)

I was running into the same issue. Developed something that shouldn't be used, but fixes my problem:
https://github.com/fancy-whale/KuredHorn

I gave up on trying to fix the pdb and removing it because I was scared LongHorn would fix this at some point. Hence the bad workaround. This is not supported behavior and shouldn't be used, but let's just say I sleep better at night with k3s and kured not pestering anymore.

**Comment by pat-s** [comment]
@gcarrarom My current workaround is to use the `force-reboot` flag with `kured`. I haven't yet inspected whether this drains the node rebooting or does a hard reboot, but for the moment I haven't seen any bad consequences in relation to this flag.

**Comment by SPodjasek** [comment]
I've run to similar issue on Longhorn v1.5.5 when doing Kubernetes cluster upgrade, first node went fine, second one blocked not being able to evict instance manager. Still got three more to go - hope it doesn't occur any more.

Detailed description of my case here: https://github.com/longhorn/longhorn/issues/8587#issuecomment-2808848622

**Comment by ivan-piskunov-betpawa** [comment]
Another way how to avoid that:
1) Disable node from UI (not mandatory, but just in case)
2) drain with flag `--disable-eviction` (Use when you know what are you doing, because it applies to all `pdb`)

**Comment by mateuszdrab** [comment]
This is actually still an issue for me, the node that is stuck right now is actually disabled in the Longhorn Web UI, has no replicas, yet system-upgrade-controller is unable to drain it due to the instance-manager.
I'm running Longhorn 1.9.2 - I'd really like to find a way to do this. 
Kured seems to be able to manage it with the force reboot option.
